drivers/flash/flash_gd32_v1.c

```diff
@@ -0,0 +1,198 @@
+ * Flash size smaller than 512KB, and page size is same.
+ */
+#define FLASH_GD32_BANK0_PAGE_SIZE DT_PROP(GD32_NV_FLASH_NODE, bank0_page_size)
+
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+static const struct flash_pages_layout gd32_fmc_v1_layout[] = {
+	{
+	.pages_size = FLASH_GD32_BANK0_PAGE_SIZE,
+	.pages_count = FLASH_GD32_SIZE / FLASH_GD32_BANK0_PAGE_SIZE
+	}
+};
+#endif
+
+static inline void flash_gd32_fmc_v1_unlock(void)
+{
+	FMC_KEY = UNLOCK_KEY0;
+	FMC_KEY = UNLOCK_KEY1;
+}
+
+static inline void flash_gd32_fmc_v1_lock(void)
+{
+	FMC_CTL |= FMC_CTL_LK;
+}
+
+static int flash_gd32_fmc_v1_wait_idle(void)
+{
+	const int64_t expired_time = k_uptime_get() + FLASH_GD32_TIMEOUT;
+
+	while (FMC_STAT & FMC_STAT_BUSY) {
+		if (k_uptime_get() > expired_time) {
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+bool flash_gd32_valid_range(off_t offset, uint32_t len, bool write)
+{
+	if ((offset > FLASH_GD32_SIZE) ||
+	    ((offset + len) > FLASH_GD32_SIZE)) {
+		return false;
+	}
+
+	if (write) {
+		/* Check offset and len is flash_prg_t aligned. */
+		if ((offset % sizeof(flash_prg_t)) ||
+		    (len % sizeof(flash_prg_t))) {
+			return false;
+		}
+
+#ifdef FLASH_GD32_FMC_WORK_ALIGNED
+		/* Check offset and len is word aligned. */
+		if ((offset % sizeof(uint32_t)) ||
+		    (len % sizeof(uint32_t))) {
+			return false;
+		}
+#endif
+
+	} else {
+		if ((offset % FLASH_GD32_BANK0_PAGE_SIZE) ||
+		    (len % FLASH_GD32_BANK0_PAGE_SIZE)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+int flash_gd32_write_range(off_t offset, const void *data, size_t len)
+{
+	flash_prg_t *prg_flash = (flash_prg_t *)((uint8_t *)FLASH_GD32_ADDR + offset);
+	flash_prg_t *prg_data = (flash_prg_t *)data;
+	int ret = 0;
+
+	flash_gd32_fmc_v1_unlock();
+
+	if (FMC_STAT & FMC_STAT_BUSY) {
+		return -EBUSY;
+	}
+
+	FMC_CTL |= FMC_CTL_PG;
+
+	for (size_t i = 0U; i < (len / sizeof(flash_prg_t)); i++) {
+		*prg_flash++ = *prg_data++;
+	}
+
+	ret = flash_gd32_fmc_v1_wait_idle();
+	if (ret < 0) {
+		goto expired_out;
+	}
+
+	if (FMC_STAT & FLASH_GD32_FMC_V1_WRITE_ERR) {
+		ret = -EIO;
+		FMC_STAT |= FLASH_GD32_FMC_V1_WRITE_ERR;
+		LOG_ERR("FMC programming failed");
+	}
+
+expired_out:
+	FMC_CTL &= ~FMC_CTL_PG;
+
+	flash_gd32_fmc_v1_lock();
+
+	return ret;
+}
+
+static int flash_gd32_fmc_v1_page_erase(uint32_t page_addr)
+{
+	int ret = 0;
+
+	flash_gd32_fmc_v1_unlock();
+
+	if (FMC_STAT & FMC_STAT_BUSY) {
+		return -EBUSY;
+	}
+
+	FMC_CTL |= FMC_CTL_PER;
+
+	FMC_ADDR = page_addr;
+
+	FMC_CTL |= FMC_CTL_START;
+
+	ret = flash_gd32_fmc_v1_wait_idle();
+	if (ret < 0) {
+		goto expired_out;
+	}
+
+	if (FMC_STAT & FLASH_GD32_FMC_V1_ERASE_ERR) {
+		ret = -EIO;
+		FMC_STAT |= FLASH_GD32_FMC_V1_ERASE_ERR;
+		LOG_ERR("FMC page %u erase failed", page_addr);
+	}
+
+expired_out:
+	FMC_CTL &= ~FMC_CTL_PER;
+
+	flash_gd32_fmc_v1_lock();
+
+	return ret;
+}
+
+int flash_gd32_erase_block(off_t offset, size_t size)
+{
+	uint32_t page_addr = FLASH_GD32_ADDR + offset;
+	int ret = 0;
+
+	while (size) {
```

Comment:
What if `size` gets below 0 because of setting/passing wrong values here?

Reply:
Thanks for point out this, @soburi first mention this for me, but I'm not sure their have any requirement for this. Anyway, change to use this type.